File: crates/astrcode-extension-contract/src/protocol.rs
use std::{
    collections::BTreeSet,
    fmt::{self, Display},
};

use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;

use crate::{WireErrorCode, manifest::InitializeManifest};

pub const S5R_VERSION: &str = "3.0";
pub const S5R_STACK: &str = "astrcode";
pub const WIRE_CODEC_JSON: &str = "json";

pub const FEATURE_NESTED_INVOKE_V1: &str = "nested_invoke_v1";
pub const FEATURE_MODEL_STREAM_V1: &str = "model_stream_v1";
pub const FEATURE_CUSTOM_EVENT_V1: &str = "custom_event_v1";
pub const CAP_HANDLER_INVOKE: &str = "handler.invoke";
pub const CAP_RUNTIME_PING: &str = "s5r.runtime.ping";
pub const CONFORMANCE_UNARY: &str = "s5r.conformance.unary";
pub const CONFORMANCE_STREAM: &str = "s5r.conformance.stream";
pub const CONFORMANCE_NESTED: &str = "s5r.conformance.nested";
pub const CONFORMANCE_WAIT_FOR_CANCEL: &str = "s5r.conformance.wait_for_cancel";
pub const CONFORMANCE_UNKNOWN_ERROR: &str = "s5r.conformance.unknown_error";
pub const CONFORMANCE_HOST_ECHO: &str = "s5r.conformance.host_echo";

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
#[serde(transparent)]
pub struct HandlerId(String);

impl HandlerId {
    pub fn new(extension_id: &str, kind: HandlerKind, name: &str) -> Result<Self, String> {
        let wire = format!("{extension_id}:{}:{name}", kind.as_str());
        Self::parse(&wire).ok_or_else(|| format!("invalid handler id {wire:?}"))
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn parse(wire: &str) -> Option<Self> {
        Self::split(wire).map(|_| Self(wire.to_owned()))
    }

    pub fn split(wire: &str) -> Option<(&str, HandlerKind, &str)> {
        let (extension_id, remainder) = wire.split_once(':')?;
        let (kind, name) = remainder.split_once(':')?;
        let kind = HandlerKind::parse(kind)?;
        if extension_id.is_empty() || name.is_empty() {
            return None;
        }
        Some((extension_id, kind, name))
    }

    pub fn parts(&self) -> (&str, HandlerKind, &str) {
        Self::split(&self.0).expect("HandlerId constructors preserve the parsed invariant")
    }
}

impl From<HandlerId> for String {
    fn from(id: HandlerId) -> Self {
        id.0
    }
}

impl Display for HandlerId {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl<'de> Deserialize<'de> for HandlerId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let wire = String::deserialize(deserializer)?;
        Self::parse(&wire)
            .ok_or_else(|| serde::de::Error::custom(format!("invalid handler id {wire:?}")))
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HandlerKind {
    Tool,
    Hook,
    Command,
    Http,
    Event,
}

impl HandlerKind {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Tool => "tool",
            Self::Hook => "hook",
            Self::Command => "command",
            Self::Http => "http",
            Self::Event => "event",
        }
    }

    pub fn parse(kind: &str) -> Option<Self> {
        Some(match kind {
            "tool" => Self::Tool,
            "hook" => Self::Hook,
            "command" => Self::Command,
            "http" => Self::Http,
            "event" => Self::Event,
            _ => return None,
        })
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
#[serde(transparent)]
pub struct FeatureName(String);

impl FeatureName {
    pub fn parse(value: impl Into<String>) -> Result<Self, String> {
        let value = value.into();
        if valid_feature_name(&value) {
            Ok(Self(value))
        } else {
            Err(format!("invalid feature name {value:?}"))
        }
    }

    pub fn nested_invoke_v1() -> Self {
        Self(FEATURE_NESTED_INVOKE_V1.into())
    }

    pub fn model_stream_v1() -> Self {
        Self(FEATURE_MODEL_STREAM_V1.into())
    }

    pub fn custom_event_v1() -> Self {
        Self(FEATURE_CUSTOM_EVENT_V1.into())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl<'de> Deserialize<'de> for FeatureName {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Self::parse(String::deserialize(deserializer)?).map_err(serde::de::Error::custom)
    }
}

impl Display for FeatureName {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.0)
    }
}

fn valid_feature_name(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'_')
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct PeerInfo {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
}

impl PeerInfo {
    pub fn validate(&self) -> Result<(), ErrorPayload> {
        if self.name.is_empty() {
            return Err(ErrorPayload::new(
                WireErrorCode::InvalidRequest,
                "peer name must not be empty",
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct HandlerInvokeRequest {
    pub handler_id: HandlerId,
    #[serde(default)]
    pub event: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct InitializeMsg {
    pub id: String,
    pub protocol_version: String,
    pub host: PeerInfo,
    pub extension_id: String,
    #[serde(default)]
    pub supported_features: Vec<FeatureName>,
    #[serde(default)]
    pub required_features: Vec<FeatureName>,
    #[serde(default)]
    pub host_operations: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct InitializeOutput {
    pub worker: PeerInfo,
    pub protocol_version: String,
    #[serde(default)]
    pub supported_features: Vec<FeatureName>,
    #[serde(default)]
    pub required_features: Vec<FeatureName>,
    pub negotiated_features: Vec<FeatureName>,
    pub manifest: InitializeManifest,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivateMsg {
    pub id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ActivateOutput {}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ResultKind {
    Initialize,
    Activate,
    Invoke,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case", deny_unknown_fields)]
pub enum ResultMsg {
    Success {
        id: String,
        kind: ResultKind,
        output: Value,
    },
    Failure {
        id: String,
        kind: ResultKind,
        error: ErrorPayload,
    },
}

impl ResultMsg {
    pub fn success(id: String, kind: ResultKind, output: Value) -> Self {
        Self::Success { id, kind, output }
    }

    pub fn failure(id: String, kind: ResultKind, error: ErrorPayload) -> Self {
        Self::Failure { id, kind, error }
    }

    pub fn id(&self) -> &str {
        match self {
            Self::Success { id, .. } | Self::Failure { id, .. } => id,
        }
    }

    pub const fn kind(&self) -> ResultKind {
        match self {
            Self::Success { kind, .. } | Self::Failure { kind, .. } => *kind,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct InvokeMsg {
    pub id: String,
    pub operation: String,
    #[serde(default)]
    pub input: Value,
    #[serde(default)]
    pub stream: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parent_invoke_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct StreamMsg {
    pub id: String,
    pub event: ModelStreamEvent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum ModelStreamEvent {
    Started,
    Retrying {
        attempt: u32,
        delay_ms: u64,
    },
    Recovered {
        attempt: u32,
    },
    ContentDelta {
        content: String,
    },
    ThinkingDelta {
        content: String,
    },
    ToolCallStart {
        tool_call_id: String,
        name: String,
        arguments: String,
    },
    ToolCallDelta {
        tool_call_id: String,
        delta: String,
    },
    ToolCallCompleted {
        tool_call_id: String,
    },
    Usage {
        input_tokens: u64,
        output_tokens: u64,
    },
    Completed {
        output: Value,
    },
    Failed {
        error: ErrorPayload,
    },
}

impl ModelStreamEvent {
    pub const fn is_terminal(&self) -> bool {
        matches!(self, Self::Completed { .. } | Self::Failed { .. })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CancelMsg {
    pub id: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum WireMessage {
    Initialize(InitializeMsg),
    Activate(ActivateMsg),
    Result(ResultMsg),
    Invoke(InvokeMsg),
    Stream(StreamMsg),
    Cancel(CancelMsg),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct ErrorPayload {
    pub code: String,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hint: Option<String>,
    #[serde(default)]
    pub retryable: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub details: Option<Value>,
}

impl ErrorPayload {
    pub fn new(code: WireErrorCode, message: impl Into<String>) -> Self {
        Self {
            code: code.as_str().into(),
            message: message.into(),
            hint: None,
            retryable: false,
            details: None,
        }
    }

    pub fn code_enum(&self) -> Option<WireErrorCode> {
        WireErrorCode::parse(&self.code)
    }

    pub fn with_hint(mut self, hint: impl Into<String>) -> Self {
        self.hint = Some(hint.into());
        self
    }

    pub const fn retryable(mut self, retryable: bool) -> Self {
        self.retryable = retryable;
        self
    }
}

impl fmt::Display for ErrorPayload {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}: {}", self.code, self.message)
    }
}

pub fn encode_wire_message(message: &WireMessage) -> Result<Vec<u8>, ProtocolError> {
    serde_json::to_vec(message).map_err(ProtocolError::Encode)
}

pub fn parse_wire_message(payload: &[u8]) -> Result<WireMessage, ProtocolError> {
    serde_json::from_slice(payload).map_err(ProtocolError::Decode)
}

pub fn negotiate_features(
    local_supported: &BTreeSet<FeatureName>,
    remote_supported: &[FeatureName],
    remote_required: &[FeatureName],
) -> Result<BTreeSet<FeatureName>, ErrorPayload> {
    ensure_unique(remote_supported, "supported_features")?;
    ensure_unique(remote_required, "required_features")?;

    let remote_supported: BTreeSet<_> = remote_supported.iter().cloned().collect();
    for feature in remote_required {
        if !remote_supported.contains(feature) {
            return Err(ErrorPayload::new(
                WireErrorCode::InvalidRequest,
                format!("required feature {feature} is not declared as supported"),
            ));
        }
        if !local_supported.contains(feature) {
            return Err(ErrorPayload::new(
                WireErrorCode::UnsupportedFeature,
                format!("required feature {feature} is not supported by this peer"),
            ));
        }
    }
    Ok(local_supported
        .intersection(&remote_supported)
        .cloned()
        .collect())
}

fn ensure_unique(features: &[FeatureName], field: &str) -> Result<(), ErrorPayload> {
    let unique: BTreeSet<_> = features.iter().collect();
    if unique.len() == features.len() {
        Ok(())
    } else {
        Err(ErrorPayload::new(
            WireErrorCode::InvalidRequest,
            format!("{field} contains duplicate values"),
        ))
    }
}

#[derive(Debug, thiserror::Error)]
pub enum ProtocolError {
    #[error("encode S5R message: {0}")]
    Encode(serde_json::Error),
    #[error("decode S5R message: {0}")]
    Decode(serde_json::Error),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn feature_negotiation_is_sorted_strict_and_forward_compatible() {
        let local = BTreeSet::from([
            FeatureName::model_stream_v1(),
            FeatureName::nested_invoke_v1(),
        ]);
        let unknown = FeatureName::parse("future_feature_v1").unwrap();
        let negotiated = negotiate_features(
            &local,
            &[unknown, FeatureName::nested_invoke_v1()],
            &[FeatureName::nested_invoke_v1()],
        )
        .unwrap();
        assert_eq!(
            negotiated,
            BTreeSet::from([FeatureName::nested_invoke_v1()])
        );

        let error = negotiate_features(
            &local,
            &[FeatureName::custom_event_v1()],
            &[FeatureName::custom_event_v1()],
        )
        .unwrap_err();
        assert_eq!(error.code_enum(), Some(WireErrorCode::UnsupportedFeature));
    }

    #[test]
    fn envelope_and_nested_payloads_reject_unknown_fields() {
        let cases = [
            br#"{"type":"initialize","id":"1","protocol_version":"3.0","host":{"name":"host"},"extension_id":"worker","host_operations":[],"unknown":true}"#.as_slice(),
            br#"{"type":"initialize","id":"1","protocol_version":"3.0","host":{"name":"host","unknown":true},"extension_id":"worker","host_operations":[]}"#.as_slice(),
            br#"{"type":"activate","id":"1","unknown":true}"#.as_slice(),
            br#"{"type":"cancel","id":"1","reason":"reload","unknown":true}"#.as_slice(),
            br#"{"type":"stream","id":"1","event":{"type":"content_delta","content":"delta","unknown":true}}"#.as_slice(),
        ];
        for payload in cases {
            assert!(parse_wire_message(payload).is_err());
        }

        assert!(
            serde_json::from_value::<InitializeOutput>(serde_json::json!({
                "worker": {"name": "worker"},
                "protocol_version": "3.0",
                "negotiated_features": [],
                "manifest": {"unknown": true}
            }))
            .is_err()
        );
        assert!(
            serde_json::from_value::<ActivateOutput>(serde_json::json!({"unknown": true})).is_err()
        );
    }

    #[test]
    fn unknown_error_codes_are_preserved_losslessly() {
        let error: ErrorPayload = serde_json::from_value(serde_json::json!({
            "code": "future_remote_failure",
            "message": "new peer error"
        }))
        .unwrap();
        assert_eq!(error.code, "future_remote_failure");
        assert_eq!(error.code_enum(), None);
    }
}

File: sdks/python/src/s5r/protocol.py
"""S5R 3.0 wire messages: strict decode, encode helpers, feature negotiation.

Mirrors `astrcode_extension_sdk::wire::protocol`. Envelopes and nested payloads
are parsed strictly: unknown fields, unknown `type` tags, and malformed values
raise `ProtocolError`. All field names are snake_case.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Mapping, Union

from .errors import ErrorPayload, ProtocolError, S5rError, WireErrorCode

S5R_VERSION = "3.0"

FEATURE_NESTED_INVOKE_V1 = "nested_invoke_v1"
FEATURE_MODEL_STREAM_V1 = "model_stream_v1"
FEATURE_CUSTOM_EVENT_V1 = "custom_event_v1"

CAP_HANDLER_INVOKE = "handler.invoke"
CAP_RUNTIME_PING = "s5r.runtime.ping"

CONFORMANCE_UNARY = "s5r.conformance.unary"
CONFORMANCE_STREAM = "s5r.conformance.stream"
CONFORMANCE_NESTED = "s5r.conformance.nested"
CONFORMANCE_WAIT_FOR_CANCEL = "s5r.conformance.wait_for_cancel"
CONFORMANCE_UNKNOWN_ERROR = "s5r.conformance.unknown_error"
CONFORMANCE_HOST_ECHO = "s5r.conformance.host_echo"

RESULT_KINDS = frozenset({"initialize", "activate", "invoke"})

_MESSAGE_FIELDS: dict[str, frozenset[str]] = {
    "initialize": frozenset(
        {
            "type",
            "id",
            "protocol_version",
            "host",
            "extension_id",
            "supported_features",
            "required_features",
            "host_operations",
        }
    ),
    "activate": frozenset({"type", "id", "config"}),
    "result": frozenset({"type", "status", "id", "kind", "output", "error"}),
    "invoke": frozenset(
        {"type", "id", "operation", "input", "stream", "parent_invoke_id"}
    ),
    "stream": frozenset({"type", "id", "event"}),
    "cancel": frozenset({"type", "id", "reason"}),
}

_STREAM_EVENT_FIELDS: dict[str, frozenset[str]] = {
    "started": frozenset({"type"}),
    "retrying": frozenset({"type", "attempt", "delay_ms"}),
    "recovered": frozenset({"type", "attempt"}),
    "content_delta": frozenset({"type", "content"}),
    "thinking_delta": frozenset({"type", "content"}),
    "tool_call_start": frozenset({"type", "tool_call_id", "name", "arguments"}),
    "tool_call_delta": frozenset({"type", "tool_call_id", "delta"}),
    "tool_call_completed": frozenset({"type", "tool_call_id"}),
    "usage": frozenset({"type", "input_tokens", "output_tokens"}),
    "completed": frozenset({"type", "output"}),
    "failed": frozenset({"type", "error"}),
}

TERMINAL_STREAM_EVENTS = frozenset({"completed", "failed"})


@dataclass
class InitializeMsg:
    id: str
    protocol_version: str
    host_name: str
    host_version: str | None
    extension_id: str
    supported_features: list[str] = field(default_factory=list)
    required_features: list[str] = field(default_factory=list)
    host_operations: list[str] = field(default_factory=list)


@dataclass
class ActivateMsg:
    id: str
    config: Any


@dataclass
class ResultMsg:
    id: str
    kind: str
    output: Any = None
    error: ErrorPayload | None = None

    @property
    def is_success(self) -> bool:
        return self.error is None


@dataclass
class InvokeMsg:
    id: str
    operation: str
    input: Any = None
    stream: bool = False
    parent_invoke_id: str | None = None


@dataclass
class StreamMsg:
    id: str
    event: dict[str, Any]


@dataclass
class CancelMsg:
    id: str
    reason: str


Message = Union[InitializeMsg, ActivateMsg, ResultMsg, InvokeMsg, StreamMsg, CancelMsg]


def encode_message(message: Mapping[str, Any]) -> bytes:
    return json.dumps(message, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def result_success(request_id: str, kind: str, output: Any) -> dict[str, Any]:
    return {
        "type": "result",
        "status": "success",
        "id": request_id,
        "kind": kind,
        "output": output,
    }


def result_failure(request_id: str, kind: str, error: ErrorPayload) -> dict[str, Any]:
    return {
        "type": "result",
        "status": "failure",
        "id": request_id,
        "kind": kind,
        "error": error.to_json(),
    }


def invoke_message(
    request_id: str,
    operation: str,
    input: Any,
    *,
    stream: bool = False,
    parent_invoke_id: str | None = None,
) -> dict[str, Any]:
    message: dict[str, Any] = {
        "type": "invoke",
        "id": request_id,
        "operation": operation,
        "input": input,
        "stream": stream,
    }
    if parent_invoke_id is not None:
        message["parent_invoke_id"] = parent_invoke_id
    return message


def stream_message(request_id: str, event: Mapping[str, Any]) -> dict[str, Any]:
    return {"type": "stream", "id": request_id, "event": dict(event)}


def cancel_message(request_id: str, reason: str) -> dict[str, Any]:
    return {"type": "cancel", "id": request_id, "reason": reason}


def decode_message(payload: bytes) -> Message:
    try:
        value = json.loads(payload)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ProtocolError(f"decode S5R message: {error}") from error
    if not isinstance(value, dict):
        raise ProtocolError("S5R envelope must be a JSON object")
    msg_type = value.get("type")
    allowed = _MESSAGE_FIELDS.get(msg_type if isinstance(msg_type, str) else "")
    if allowed is None:
        raise ProtocolError(f"unknown S5R message type {msg_type!r}")
    unknown = set(value) - allowed
    if unknown:
        raise ProtocolError(f"{msg_type} message has unknown fields: {sorted(unknown)}")
    if msg_type == "initialize":
        return _decode_initialize(value)
    if msg_type == "activate":
        return ActivateMsg(id=_required_str(value, "id", "activate"), config=value.get("config"))
    if msg_type == "result":
        return _decode_result(value)
    if msg_type == "invoke":
        return _decode_invoke(value)
    if msg_type == "stream":
        request_id = _required_str(value, "id", "stream")
        event = value.get("event")
        validate_stream_event(event)
        return StreamMsg(id=request_id, event=event)
    return CancelMsg(
        id=_required_str(value, "id", "cancel"),
        reason=_required_str(value, "reason", "cancel"),
    )


def validate_stream_event(event: Any) -> None:
    if not isinstance(event, dict):
        raise ProtocolError("stream event must be an object")
    event_type = event.get("type")
    allowed = _STREAM_EVENT_FIELDS.get(event_type if isinstance(event_type, str) else "")
    if allowed is None:
        raise ProtocolError(f"unknown stream event type {event_type!r}")
    unknown = set(event) - allowed
    if unknown:
        raise ProtocolError(
            f"{event_type} stream event has unknown fields: {sorted(unknown)}"
        )
    missing = allowed - {"type"} - set(event)
    if missing:
        raise ProtocolError(
            f"{event_type} stream event is missing fields: {sorted(missing)}"
        )
    if event_type == "failed":
        ErrorPayload.from_json(event["error"])


def valid_feature_name(value: str) -> bool:
    return (
        0 < len(value) <= 64
        and all(c.isascii() and (c.islower() or c.isdigit()) or c == "_" for c in value)
    )


def negotiate_features(
    local_supported: set[str],
    remote_supported: list[str],
    remote_required: list[str],
) -> set[str]:
    """Intersection semantics shared by both peers; raises `S5rError` on failure."""
    if len(set(remote_supported)) != len(remote_supported):
        raise S5rError.of(
            WireErrorCode.INVALID_REQUEST, "supported_features contains duplicate values"
        )
    if len(set(remote_required)) != len(remote_required):
        raise S5rError.of(
            WireErrorCode.INVALID_REQUEST, "required_features contains duplicate values"
        )
    remote_supported_set = set(remote_supported)
    for feature in remote_required:
        if feature not in remote_supported_set:
            raise S5rError.of(
                WireErrorCode.INVALID_REQUEST,
                f"required feature {feature} is not declared as supported",
            )
        if feature not in local_supported:
            raise S5rError.of(
                WireErrorCode.UNSUPPORTED_FEATURE,
                f"required feature {feature} is not supported by this peer",
            )
    return local_supported & remote_supported_set


def _required_str(value: Mapping[str, Any], key: str, what: str) -> str:
    field_value = value.get(key)
    if not isinstance(field_value, str):
        raise ProtocolError(f"{what} message requires string field {key!r}")
    return field_value


def _decode_initialize(value: Mapping[str, Any]) -> InitializeMsg:
    host = value.get("host")
    if not isinstance(host, dict):
        raise ProtocolError("initialize message requires object field 'host'")
    unknown_host = set(host) - {"name", "version"}
    if unknown_host:
        raise ProtocolError(f"host peer info has unknown fields: {sorted(unknown_host)}")
    host_name = host.get("name")
    host_version = host.get("version")
    if not isinstance(host_name, str):
        raise ProtocolError("host peer info requires string field 'name'")
    if host_version is not None and not isinstance(host_version, str):
        raise ProtocolError("host peer info version must be a string")
    return InitializeMsg(
        id=_required_str(value, "id", "initialize"),
        protocol_version=_required_str(value, "protocol_version", "initialize"),
        host_name=host_name,
        host_version=host_version,
        extension_id=_required_str(value, "extension_id", "initialize"),
        supported_features=_decode_features(value.get("supported_features", [])),
        required_features=_decode_features(value.get("required_features", [])),
        host_operations=_decode_host_operations(value.get("host_operations", [])),
    )


def _decode_features(value: Any) -> list[str]:
    if not isinstance(value, list):
        raise ProtocolError("feature lists must be arrays")
    for feature in value:
        if not isinstance(feature, str) or not valid_feature_name(feature):
            raise ProtocolError(f"invalid feature name {feature!r}")
    return list(value)


def _decode_host_operations(value: Any) -> list[str]:
    if not isinstance(value, list) or not all(isinstance(op, str) for op in value):
        raise ProtocolError("host_operations must be an array of strings")
    return list(value)


def _decode_result(value: Mapping[str, Any]) -> ResultMsg:
    request_id = _required_str(value, "id", "result")
    kind = _required_str(value, "kind", "result")
    if kind not in RESULT_KINDS:
        raise ProtocolError(f"unknown result kind {kind!r}")
    status = value.get("status")
    if status == "success":
        if "output" not in value:
            raise ProtocolError("successful result must carry output")
        if "error" in value:
            raise ProtocolError("successful result must not carry error")
        return ResultMsg(id=request_id, kind=kind, output=value["output"])
    if status == "failure":
        if "error" not in value:
            raise ProtocolError("failed result must carry error")
        if "output" in value:
            raise ProtocolError("failed result must not carry output")
        return ResultMsg(
            id=request_id, kind=kind, error=ErrorPayload.from_json(value["error"])
        )
    raise ProtocolError(f"unknown result status {status!r}")


def _decode_invoke(value: Mapping[str, Any]) -> InvokeMsg:
    stream = value.get("stream", False)
    parent = value.get("parent_invoke_id")
    if not isinstance(stream, bool):
        raise ProtocolError("invoke stream flag must be a boolean")
    if parent is not None and not isinstance(parent, str):
        raise ProtocolError("invoke parent_invoke_id must be a string")
    return InvokeMsg(
        id=_required_str(value, "id", "invoke"),
        operation=_required_str(value, "operation", "invoke"),
        input=value.get("input"),
        stream=stream,
        parent_invoke_id=parent,
    )
